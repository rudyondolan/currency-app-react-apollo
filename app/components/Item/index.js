import React from 'react';
import { FormattedNumber } from 'react-intl';
import { Segment, Grid, Header, Icon } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { find, omit, isEqual } from 'lodash';

const IconWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

const StyledIcon = styled(Icon)`
  &&&&&& {
    margin-left: auto;
    margin-right: 0;
    cursor: pointer;
    background-color: #d0d0d0 !important;
    &:hover {
      background-color: red !important;
    }
  }
`;

class Item extends React.PureComponent {
  constructor(props) {
    super(props);
    this.handleClose = this.handleClose.bind(this);
  }

  handleClose() {
    const {
      onClose,
      currencyData: { id },
      totalCurrency,
      onToggleForm,
      onToggleButton,
      onResetCurrency,
      isDisplayForm,
    } = this.props;

    if (totalCurrency <= 1) {
      if (!isDisplayForm) {
        onToggleForm();
        onToggleButton();
      }
      onResetCurrency();
    }

    onClose(id);
  }

  determineAmount(amount) {
    const {
      data: { rates },
      currencyData: {
        data: { convertedCurrency },
      },
    } = this.props;

    const findData = find(rates, (value, key) => key === convertedCurrency);
    const extractData = omit(findData, ['__typename']);
    const { amount: originAmount } = extractData;

    if (!isEqual(amount, originAmount)) {
      return originAmount;
    }

    return amount;
  }

  determineBase(base) {
    const {
      data: { rates },
      currencyData: {
        data: { convertedCurrency },
      },
    } = this.props;

    const findData = find(rates, (value, key) => key === convertedCurrency);
    const extractData = omit(findData, ['__typename']);
    const { base: originBase } = extractData;

    if (!isEqual(base, originBase)) {
      return originBase;
    }

    return base;
  }

  determineCurrency(currency) {
    const {
      data: { base: originCurrency },
    } = this.props;
    if (!isEqual(currency, originCurrency)) {
      return originCurrency;
    }

    return currency;
  }

  renderConvertedCurrency(convertedCurrency) {
    return <Header as="h3">{convertedCurrency}</Header>;
  }

  renderAmount(amount) {
    const currencyAmount = this.determineAmount(amount);
    return (
      <Header as="h3">
        <FormattedNumber value={currencyAmount} minimumFractionDigits={2} />
      </Header>
    );
  }

  renderText(text) {
    return <Header as="h4">{text}</Header>;
  }

  renderBase(currency, convertedCurrency, base) {
    const currencyCode = this.determineCurrency(currency);
    const currencyBase = this.determineBase(base);

    return (
      <Header as="h5">
        <i>
          {`1 ${currencyCode} = ${convertedCurrency} `}
          <FormattedNumber value={currencyBase} minimumFractionDigits={2} />
        </i>
      </Header>
    );
  }

  render() {
    const {
      loading,
      currencyData: {
        currency,
        data: { convertedCurrency, amount, text, base },
      },
    } = this.props;

    return (
      <Segment loading={loading}>
        <IconWrapper>
          <StyledIcon
            size="small"
            onClick={this.handleClose}
            circular
            name="close"
            inverted
          />
        </IconWrapper>
        <Grid columns={2} celled style={{ marginBottom: '0px' }}>
          <Grid.Row>
            <Grid.Column>
              {!loading && this.renderConvertedCurrency(convertedCurrency)}
            </Grid.Column>
            <Grid.Column>{!loading && this.renderAmount(amount)}</Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <Grid.Column>{!loading && this.renderText(text)}</Grid.Column>
          </Grid.Row>
          <Grid.Row>
            <Grid.Column>
              {!loading && this.renderBase(currency, convertedCurrency, base)}
            </Grid.Column>
          </Grid.Row>
        </Grid>
      </Segment>
    );
  }
}

Item.propTypes = {
  totalCurrency: PropTypes.number,
  currencyData: PropTypes.object,
  loading: PropTypes.bool,
  data: PropTypes.object,
  isDisplayForm: PropTypes.bool,
  onClose: PropTypes.func,
  onToggleForm: PropTypes.func,
  onToggleButton: PropTypes.func,
  onResetCurrency: PropTypes.func,
};

export default Item;
