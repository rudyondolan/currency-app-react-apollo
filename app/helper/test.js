import React from 'react';
import { IntlProvider, intlShape } from 'react-intl';
import { mount, shallow } from 'enzyme';

const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();

function nodeWithIntlProp(node) {
  return React.cloneElement(node, { intl });
}

export function shallowWithIntl(node, { context, ...additionalOptions } = {}) {
  return shallow(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl }),
    ...additionalOptions,
  });
}

export function mountWithIntl(
  node,
  { context, childContextTypes, ...additionalOptions } = {},
) {
  return mount(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl }),
    childContextTypes: Object.assign(
      {},
      { intl: intlShape },
      childContextTypes,
    ),
    ...additionalOptions,
  });
}
