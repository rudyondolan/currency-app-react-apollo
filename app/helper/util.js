export function toFloat(value) {
  if (typeof value === 'string') {
    const fVal = parseFloat(value);
    if (typeof fVal === 'number' && !isNaN(fVal)) {
      return fVal;
    }

    return 1;
  }

  return value;
}

