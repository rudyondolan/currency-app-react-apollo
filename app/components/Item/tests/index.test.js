import React from 'react';
import { Icon } from 'semantic-ui-react';
import { mountWithIntl } from 'helper/test';
import Item from '../index';

describe('<Item />', () => {
  test('should match the snapshot', () => {
    const wrapper = renderItem();
    expect(wrapper).toMatchSnapshot('before update');

    const props = {
      data: {
        base: 'USD',
        date: '2018-07-20',
        rates: {
          AUD: {
            amount: 13.554413025,
            base: 1.3554413025,
            text: 'Australia Dollar',
          },
        },
      },
    };

    wrapper.setProps(props);
    expect(wrapper).toMatchSnapshot('after update');
  });

  test('should call click event when close the item', () => {
    const onClose = jest.fn();
    const props = { onClose };
    const wrapper = renderItem(props);
    wrapper.find(Icon).simulate('click');
    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

function renderItem(props = {}) {
  const propsToUse = {
    totalCurrency: 1,
    currencyData: {
      id: 0,
      currency: 'USD',
      data: {
        amount: 1.3554413025,
        base: 1.3554413025,
        convertedCurrency: 'AUD',
        text: 'Australia Dollar',
      },
    },
    loading: false,
    isDisplayForm: false,
    data: {},
    onClose() {},
    onToggleForm() {},
    onToggleButton() {},
    onResetCurrency() {},
    ...props,
  };

  return mountWithIntl(<Item {...propsToUse} />);
}
