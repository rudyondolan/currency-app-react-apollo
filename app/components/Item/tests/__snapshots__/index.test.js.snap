// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> should match the snapshot: after update 1`] = `
<Item
  currencyData={
    Object {
      "currency": "USD",
      "data": Object {
        "amount": 1.3554413025,
        "base": 1.3554413025,
        "convertedCurrency": "AUD",
        "text": "Australia Dollar",
      },
      "id": 0,
    }
  }
  data={
    Object {
      "base": "USD",
      "date": "2018-07-20",
      "rates": Object {
        "AUD": Object {
          "amount": 13.554413025,
          "base": 1.3554413025,
          "text": "Australia Dollar",
        },
      },
    }
  }
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
    }
  }
  isDisplayForm={false}
  loading={false}
  onClose={[Function]}
  onResetCurrency={[Function]}
  onToggleButton={[Function]}
  onToggleForm={[Function]}
  totalCurrency={1}
>
  <Segment
    loading={false}
  >
    <div
      className="ui segment"
    >
      <Item__IconWrapper>
        <div
          className="Item__IconWrapper-ktHDgb esfGgc"
        >
          <Item__StyledIcon
            circular={true}
            inverted={true}
            name="close"
            onClick={[Function]}
            size="small"
          >
            <Icon
              as="i"
              circular={true}
              className="Item__StyledIcon-kVInds ccSeKA"
              inverted={true}
              name="close"
              onClick={[Function]}
              size="small"
            >
              <i
                aria-hidden="true"
                className="close small circular inverted icon Item__StyledIcon-kVInds ccSeKA"
                onClick={[Function]}
              />
            </Icon>
          </Item__StyledIcon>
        </div>
      </Item__IconWrapper>
      <Grid
        celled={true}
        columns={2}
        style={
          Object {
            "marginBottom": "0px",
          }
        }
      >
        <div
          className="ui celled two column grid"
          style={
            Object {
              "marginBottom": "0px",
            }
          }
        >
          <GridRow>
            <div
              className="row"
            >
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h3"
                  >
                    <h3
                      className="ui header"
                    >
                      AUD
                    </h3>
                  </Header>
                </div>
              </GridColumn>
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h3"
                  >
                    <h3
                      className="ui header"
                    >
                      <FormattedNumber
                        minimumFractionDigits={2}
                        value={13.554413025}
                      >
                        <span>
                          13.554
                        </span>
                      </FormattedNumber>
                    </h3>
                  </Header>
                </div>
              </GridColumn>
            </div>
          </GridRow>
          <GridRow>
            <div
              className="row"
            >
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h4"
                  >
                    <h4
                      className="ui header"
                    >
                      Australia Dollar
                    </h4>
                  </Header>
                </div>
              </GridColumn>
            </div>
          </GridRow>
          <GridRow>
            <div
              className="row"
            >
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h5"
                  >
                    <h5
                      className="ui header"
                    >
                      <i>
                        1 USD = AUD 
                        <FormattedNumber
                          minimumFractionDigits={2}
                          value={1.3554413025}
                        >
                          <span>
                            1.355
                          </span>
                        </FormattedNumber>
                      </i>
                    </h5>
                  </Header>
                </div>
              </GridColumn>
            </div>
          </GridRow>
        </div>
      </Grid>
    </div>
  </Segment>
</Item>
`;

exports[`<Item /> should match the snapshot: before update 1`] = `
<Item
  currencyData={
    Object {
      "currency": "USD",
      "data": Object {
        "amount": 1.3554413025,
        "base": 1.3554413025,
        "convertedCurrency": "AUD",
        "text": "Australia Dollar",
      },
      "id": 0,
    }
  }
  data={Object {}}
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": Object {},
      "now": [Function],
      "textComponent": "span",
    }
  }
  isDisplayForm={false}
  loading={false}
  onClose={[Function]}
  onResetCurrency={[Function]}
  onToggleButton={[Function]}
  onToggleForm={[Function]}
  totalCurrency={1}
>
  <Segment
    loading={false}
  >
    <div
      className="ui segment"
    >
      <Item__IconWrapper>
        <div
          className="Item__IconWrapper-ktHDgb esfGgc"
        >
          <Item__StyledIcon
            circular={true}
            inverted={true}
            name="close"
            onClick={[Function]}
            size="small"
          >
            <Icon
              as="i"
              circular={true}
              className="Item__StyledIcon-kVInds ccSeKA"
              inverted={true}
              name="close"
              onClick={[Function]}
              size="small"
            >
              <i
                aria-hidden="true"
                className="close small circular inverted icon Item__StyledIcon-kVInds ccSeKA"
                onClick={[Function]}
              />
            </Icon>
          </Item__StyledIcon>
        </div>
      </Item__IconWrapper>
      <Grid
        celled={true}
        columns={2}
        style={
          Object {
            "marginBottom": "0px",
          }
        }
      >
        <div
          className="ui celled two column grid"
          style={
            Object {
              "marginBottom": "0px",
            }
          }
        >
          <GridRow>
            <div
              className="row"
            >
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h3"
                  >
                    <h3
                      className="ui header"
                    >
                      AUD
                    </h3>
                  </Header>
                </div>
              </GridColumn>
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h3"
                  >
                    <h3
                      className="ui header"
                    >
                      <FormattedNumber
                        minimumFractionDigits={2}
                      >
                        <span>
                          NaN
                        </span>
                      </FormattedNumber>
                    </h3>
                  </Header>
                </div>
              </GridColumn>
            </div>
          </GridRow>
          <GridRow>
            <div
              className="row"
            >
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h4"
                  >
                    <h4
                      className="ui header"
                    >
                      Australia Dollar
                    </h4>
                  </Header>
                </div>
              </GridColumn>
            </div>
          </GridRow>
          <GridRow>
            <div
              className="row"
            >
              <GridColumn>
                <div
                  className="column"
                >
                  <Header
                    as="h5"
                  >
                    <h5
                      className="ui header"
                    >
                      <i>
                        1 undefined = AUD 
                        <FormattedNumber
                          minimumFractionDigits={2}
                        >
                          <span>
                            NaN
                          </span>
                        </FormattedNumber>
                      </i>
                    </h5>
                  </Header>
                </div>
              </GridColumn>
            </div>
          </GridRow>
        </div>
      </Grid>
    </div>
  </Segment>
</Item>
`;
